lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -913,112 +912,182 @@ def test_get_inventory(
-
-        class MockTarFile:
-            def extractfile(self, path):
-                return None
-
-        def fake_tarfile_open(self, path):
-            return MockTarFile()
+        path = "metadata.log"
+        stdout_contents = b"[test]\nfoo=bar\n"
+
+        class MockBufferedReader(io.BufferedReader):
+            def __init__(self, contents: bytes):
+                # No effect, other than to quiet the linter
+                None if True else super().__init__(io.RawIOBase())
+                self.contents = contents
+                self.loop_count = wait_cnt
+
+            def close(self) -> None:
+                raise AssertionError(
+                    "This test doesn't expect the stream to be closed."
+                )
+
+            def peek(self, size=0) -> bytes:
+                if self.loop_count > 0:
+                    self.loop_count -= 1
+                    return b""
+                return peek_return
+
+            def read(self, _size: int = -1) -> bytes:
+                return self.contents
+
+        class MockPopen(subprocess.Popen):
+            def __init__(self, *_args, **_kwargs):
+                # No effect, other than to quiet the linter
+                None if True else super().__init__([])
+                if popen_fail:
+                    raise ValueError(
+                        "MockPopen pretending it was called with invalid arguments"
+                    )
+                self.stdout = MockBufferedReader(stdout_contents)
+                self.stderr = MockBufferedReader(stderr_contents)
+                self.returncode = None
+                self.loop_count = wait_cnt
+
+            def poll(self) -> Optional[int]:
+                if self.loop_count > 0:
+                    self.loop_count -= 1
+                    return None
+                self.returncode = poll_return
+                return poll_return
+
+            def kill(self) -> None:
+                raise AssertionError(
+                    "This test doesn't expect the stream to be closed."
+                )
+
+        def mock_shutil_which(
+            cmd: str, _mode: int = os.F_OK | os.X_OK, _path: Optional[str] = None
+        ):
+            assert cmd == "tar"
+            return tar_path
 
         with monkeypatch.context() as m:
-            m.setattr(tarfile, "open", fake_tarfile_open)
-            expected_error_msg = f"Unable to extract {path} from {tar.name}"
-            with pytest.raises(CacheExtractBadPath) as exc:
-                Tarball.extract(tar, path)
-            assert str(exc.value) == expected_error_msg
+            m.setattr(shutil, "which", mock_shutil_which)
+            m.setattr(subprocess, "Popen", MockPopen)
+
+            try:
+                got = Tarball.extract(tar, path)
+            except CacheExtractBadPath as exc:
+                assert tar_path
+                assert not popen_fail
+                assert str(exc) == f"Unable to extract {path} from {tar.name}"
```

Comment:
It seems what we're missing here is an explicit confirmation that we *expected* a `CacheExtractBadPath`, not just that we expected a tar path and for `Popen` to not fail. That is, that `CacheExtractBadPath` wasn't raised erroneously...